### PR TITLE
Mitigation/trust-audit-spencer

### PR DESCRIPTION
**Pull request by spengrah**
- [x] TRST-H-1
- [x] TRST-M-1
- [x] TRST-M-2
- [x] TRST-M-1
- [x] TRST-M-4
- [x] TRST-M-7
- [x] TRST-L-1
- [x] TRST-L-2
- [x] TRST-L-4
- [x] TRST-L-5
- [x] Other recommendations
  - [x] Documentation changes
  - [x] Cleaner code
  - [x] Emit events without state change
  - [x] Improving display of information

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -109,36 +113,44 @@ Here are a few examples of how a DAO might confer authorities and responsibiliti
 
 | Authority | How is it attached to the Hat? |
 | --- | ---- |
-| Signer on a multisig | Using the Hat's ERC1155 token as a condition for membership in an Orca Protocol Pod |
-| Admin of the DAO's Github repo | Using the Hat's ERC1155 token as a condition for access via Lit Protocol |
+| Signer on a multisig | Using the Hat's ERC1155--similar token as a condition for membership in an Orca Protocol Pod |
```
Comment (nintynick) [inline README.md:116]:
```suggestion
| Signer on a multisig | Using the Hat's ERC1155-similar token as a condition for membership in an Orca Protocol Pod |
```

File: src/Hats.sol [hunk 2]
```diff
@@ -790,43 +833,82 @@ contract Hats is IHats, ERC1155, HatsIdUtilities {
...
-        return isWearerOfHat(_user, getAdminAtLevel(_hatId, 0));
+        // if we get here, we're at the top of _hatId's local tree
+        linkedTreeAdmin = linkedTreeAdmins[getTophatDomain(_hatId)];
+        if (linkedTreeAdmin == 0) {
+            // tree is not linked
+            return isWearerOfHat(_user, getLocalAdminAtLevel(_hatId, 0));
+        } else {
+            if (isWearerOfHat(_user, linkedTreeAdmin)) return true; // user wears the linkedTreeAdmin
+
+            else return isAdminOfHat(_user, linkedTreeAdmin); // check if user is admin of linkedTreeAdmin (recursion)
```
Comment (topocount) [inline src/Hats.sol:878]: It's crazy to me that all this extra logic saves us gas relative to an extra storage read for linkages, but here we are! Great work sorting this out. I didn't realize this, on top of the `HatsIdUtils` library refactor was needed, and feel bad kicking this back to you now.

File: src/Hats.sol [hunk 3]
```diff
@@ -790,43 +833,82 @@ contract Hats is IHats, ERC1155, HatsIdUtilities {
...
+            } else {
+                // tree is linked
+                if (isWearerOfHat(_user, linkedTreeAdmin)) {
+                    return true;
+                } // user wears the treeAdmin
+                else {
+                    adminLocalHatLevel = getLocalHatLevel(linkedTreeAdmin);
+                    _hatId = linkedTreeAdmin;
+                }
+            }
```
Comment (topocount) [inline src/Hats.sol:852]: This is a nice short-circuit

**Review by topocount — APPROVED** [review]
Awesome! I have a handful of suggested changes and questions that optionally could lead to a change, but this looks fantastic

File: src/Hats.sol [hunk 4]
```diff
@@ -790,43 +833,82 @@ contract Hats is IHats, ERC1155, HatsIdUtilities {
     /// @param _hatId The id of the Hat for which the `_user` might be the admin
     /// @return bool Whether the `_user` has admin rights for the Hat
     function isAdminOfHat(address _user, uint256 _hatId) public view returns (bool) {
```
Comment (nintynick) [inline src/Hats.sol:835]: slick!

File: src/Hats.sol [hunk 5]
```diff
@@ -790,43 +833,82 @@ contract Hats is IHats, ERC1155, HatsIdUtilities {
...
+        and has the getHatStatus function.
+        But we still can't assume that the return data is a boolean, so we need to check that manuallys
+        */
+        if (success && returndata.length == 32) {
+            // check the returndata manually
+            uint256 uintReturndata = uint256(bytes32(returndata));
+            // false condition
+            if (uintReturndata == 0) {
+                active = false;
+                // true condition
```
Comment (nintynick) [inline src/Hats.sol:902]:
```suggestion
```

File: src/Hats.sol [hunk 6]
```diff
@@ -790,43 +833,82 @@ contract Hats is IHats, ERC1155, HatsIdUtilities {
...
+            } else if (uintReturndata == 1) {
```
Comment (nintynick) [inline src/Hats.sol:903]:
```suggestion
            } 
            // true condition
            else if (uintReturndata == 1) {
```

File: src/Hats.sol [hunk 7]
```diff
@@ -790,43 +833,82 @@ contract Hats is IHats, ERC1155, HatsIdUtilities {
...
-        if (success && returndata.length > 0) {
-            return abi.decode(returndata, (bool));
+    function _isActive(Hat memory _hat, uint256 _hatId) internal view returns (bool active) {
+        (bool success, bytes memory returndata) =
+            _hat.toggle.staticcall(abi.encodeWithSignature("getHatStatus(uint256)", _hatId));
+
+        /* 
+        if function call succeeds with data of length == 32, then we know the contract exists 
+        and has the getHatStatus function.
+        But we still can't assume that the return data is a boolean, so we need to check that manuallys
```
Comment (nintynick) [inline src/Hats.sol:894]:
```suggestion
        But we still can't assume that the return data is a boolean, so we need to check that manually
```

File: src/HatsIdUtilities.sol [hunk 8]
```diff
@@ -133,11 +162,11 @@ contract HatsIdUtilities is IHatsIdUtilities {
     /// @return uint256 The hat id of the resulting admin
     function getAdminAtLevel(uint256 _hatId, uint32 _level) public view returns (uint256) {
         uint256 linkedTreeAdmin = linkedTreeAdmins[getTophatDomain(_hatId)];
```
Comment (nintynick) [inline src/HatsIdUtilities.sol:164]: perhaps just linkedAdmin?

File: src/HatsIdUtilities.sol [hunk 9]
```diff
@@ -133,11 +162,11 @@ contract HatsIdUtilities is IHatsIdUtilities {
...
-        if (linkedTreeAdmin == 0) return getTreeAdminAtLevel(_hatId, _level);
+        if (linkedTreeAdmin == 0) return getLocalAdminAtLevel(_hatId, _level);
 
-        uint32 localTopHatLevel = getHatLevel(getTreeAdminAtLevel(_hatId, 0));
+        uint32 localTopHatLevel = getHatLevel(getLocalAdminAtLevel(_hatId, 0));
 
-        if (localTopHatLevel <= _level) return getTreeAdminAtLevel(_hatId, _level - localTopHatLevel);
+        if (localTopHatLevel <= _level) return getLocalAdminAtLevel(_hatId, _level - localTopHatLevel);
```
Comment (nintynick) [inline src/HatsIdUtilities.sol:169]: very nice

**Review by nintynick — CHANGES_REQUESTED** [review]
Reviewed alongside the audit report and all the changes look great. Left specific review comments, mostly editorial.

File: README.md [hunk 10]
```diff
@@ -321,21 +333,21 @@ Enabling this latter forking/exit scenario is an important protection for Hat we
...
 
-<p align="right">(<a href="#documentation-top">back to contents</a>)</p>
+<p align="right">(<a href="#table-of-contents">back to contents</a>)</p>
 
 ### Minting a Hat
 
 Only a Hat's admin can mint its token to a wearer.
 
-To mint a Hat, the Hat's max supply must not have already been reached, and the target wearer must not already wear the Hat.
+To mint a Hat, the Hat's max supply must not have already been reached, the target wearer must not already wear the Hat, and the target wearer must not be eligible for the hat.
```
[the remark this answers is not in the document]
Comment (spengrah) [reply README.md:342]: whoops, typo — good catch!

File: README.md [hunk 11]
```diff
@@ -109,36 +113,44 @@ Here are a few examples of how a DAO might confer authorities and responsibiliti
...
+Hats is designed to be highly composable -- it will work with any tool, application, or protocol that can interact with the ERC1155 interface. Further, it allows any number of such authorities or responsibilities to be attached to a single Hat, which greatly simplifies the process for DAOs of revoking those authorities as well as the process of role handoff.
 
 #### Exception: Hat Admins
 
 Hat admins are the one (very important!) exception to the rule that authorities are external to the Hats Protocol. Refer to the Admins section below for more details.
 
-<p align="right">(<a href="#documentation-top">back to contents</a>)</p>
+<p align="right">(<a href="#table-of-contents">back to contents</a>)</p>
+
+### ERC1155 Compatibility
```
[the remark this answers is not in the document]
Comment (spengrah) [reply README.md:130]: thanks!

File: src/HatsIdUtilities.sol [hunk 12]
```diff
@@ -93,36 +92,66 @@ contract HatsIdUtilities is IHatsIdUtilities {
...
 
-        return uint32(i);
+        // return uint32(i);
+    }
+
+    /// @notice Identifies the level a given hat in its local hat tree
+    /// @dev Similar to getHatLevel, but does not account for linked trees
+    /// @param _hatId the id of the hat in question
+    /// @return level The local level, from 0 to 14
+    function getLocalHatLevel(uint256 _hatId) public pure returns (uint32 level) {
```
[the remark this answers is not in the document]
Comment (spengrah) [reply src/HatsIdUtilities.sol:124]: Heh I went back and forth on this too. I was inspired by your variable names in `getAdminAtLevel`, and we use tree in a number of different contexts so I think `Local` is currently clearer all things considered.

File: src/HatsIdUtilities.sol [hunk 13]
```diff
@@ -93,36 +92,66 @@ contract HatsIdUtilities is IHatsIdUtilities {
...
-                ++i;
-            }
-        }
-        // TODO use getTopHatDomain here
-        uint256 treeAdmin = linkedTreeAdmins[uint32(_hatId >> (256 - TOPHAT_ADDRESS_SPACE))];
+        //     // should not overflow based on < MAX_LEVELS stopping condition
+        //     unchecked {
+        //         ++i;
+        //     }
+        // }
```
[the remark this answers is not in the document]
Comment (spengrah) [reply src/HatsIdUtilities.sol:109]: nope, just missed it when doing cleanup

File: src/Hats.sol [hunk 14]
```diff
@@ -299,13 +302,27 @@ contract Hats is IHats, ERC1155, HatsIdUtilities {
...
-        if (success && returndata.length > 0) {
-            newStatus = abi.decode(returndata, (bool));
+        /* 
+        if function call succeeds with data of length == 32, then we know the contract exists 
+        and has the getHatStatus function.
+        But we still can't assume that the return data is a boolean, so we need to check that manuallys
+        */
+        if (success && returndata.length == 32) {
+            // check the returndata manually
+            uint256 uintReturndata = uint256(bytes32(returndata));
```
[the remark this answers is not in the document]
Comment (spengrah) [reply src/Hats.sol:314]: actually, its ever so slightly more expensive. Changing to `abi.decode`...

File: src/Hats.sol (hunk 2 above)
Comment (spengrah) [reply src/Hats.sol:878]: Not all all! Was fun to untangle :)

File: src/Hats.sol [hunk 15]
```diff
@@ -790,43 +833,82 @@ contract Hats is IHats, ERC1155, HatsIdUtilities {
...
+            } else if (uintReturndata == 1) {
+                active = true;
+            }
+            // invalid condition
+            else {
+                active = _getHatStatus(_hat);
+            }
         } else {
-            return _getHatStatus(_hat);
+            active = _getHatStatus(_hat);
```
[the remark this answers is not in the document]
Comment (spengrah) [reply src/Hats.sol:911]: I actually did try...

```
_retrieveAndProcessToggleReturnData(Hat memory _hat, uint256 _hatId, function(Hat memory) internal view returns (bool) _fallbackFunc) internal view returns (bool)
```
But it ended up being more expensive on execution so I stuck with the repeated code, as messy as it looks.

File: src/Hats.sol [hunk 16]
```diff
@@ -17,7 +17,7 @@
 pragma solidity >=0.8.13;
 
 import { ERC1155 } from "lib/ERC1155/ERC1155.sol";
-// import "forge-std/Test.sol"; //remove after testing
+// import { console2 } from "forge-std/Test.sol"; //remove after testing
```
[the remark this answers is not in the document]
Comment (spengrah) [reply src/Hats.sol:20]: yea, will remove before deploy

File: src/Hats.sol [hunk 17]
```diff
@@ -299,13 +302,27 @@ contract Hats is IHats, ERC1155, HatsIdUtilities {
...
+        But we still can't assume that the return data is a boolean, so we need to check that manuallys
+        */
+        if (success && returndata.length == 32) {
+            // check the returndata manually
+            uint256 uintReturndata = uint256(bytes32(returndata));
+            // false condition
+            if (uintReturndata == 0) {
+                newStatus = false;
+                // true condition
+            } else if (uintReturndata == 1) {
```
[the remark this answers is not in the document]
Comment (spengrah) [reply src/Hats.sol:319]: the formatter likes it this way, what can I say 🤷

File: src/HatsIdUtilities.sol (hunk 12 above)
[the remark this answers is not in the document]
Comment (spengrah) [reply src/HatsIdUtilities.sol:124]: plus apparently @nintynick likes it 😉

**Pull request merged**